File: c/examples/fsm/03-blink/blink_fsm.c
// Copyright 2026 Jose M. Moya <jm.moya@upm.es>
// SPDX-License-Identifier: GPL-3.0-or-later

#include "blink_fsm.h"

#include <stddef.h>
#include <stdint.h>
#include <time.h>

#include "rxnet/config.h"
#include "rxnet/port.h"

typedef struct {
    rx_fsm_machine *machine;   /* needed to read machine->state in callbacks */
    gpio_num_t button_gpio;
    gpio_num_t light_gpio;
    bool latched_event;
    bool event_consumed;
    int output_enabled;
    unsigned int base_hz;
    rx_tick_t now;
    rx_tick_t next_toggle;
} blink_machine_data;

static blink_machine_data s_data[RXNET_MAX_RUNTIME_NODES];
static size_t s_count = 0;

enum {
    BLINK_STATE_OFF = 0,
    BLINK_STATE_X1 = 1,
    BLINK_STATE_X2 = 2,
};

static uint64_t half_period_us_for_state(int state, unsigned int base_hz) {
    uint64_t hz = (uint64_t)base_hz;
    uint64_t half_period_us;

    if (hz == 0u) {
        hz = 1u;
    }

    if (state == BLINK_STATE_X2) {
        hz *= 2u;
    }

    half_period_us = 500000u / hz;
    if (half_period_us == 0u) {
        half_period_us = 1u;
    }

    return half_period_us;
}

static void blink_machine_latch_inputs(rx_fsm_context *ctx, void *user) {
    blink_machine_data *data = (blink_machine_data *)user;

    (void)ctx;
    if (data == NULL) {
        return;
    }

    data->now = rx_tick_now();
    data->latched_event = app_driver_latch_button_event(data->button_gpio);
    data->event_consumed = false;
}

static int button_pressed(const rx_fsm_context *ctx, void *user) {
    const blink_machine_data *data = (const blink_machine_data *)user;

    (void)ctx;
    return data != NULL && data->latched_event;
}

static int toggle_due(const rx_fsm_context *ctx, void *user) {
    const blink_machine_data *data = (const blink_machine_data *)user;

    (void)ctx;
    return data != NULL &&
           data->machine->state != BLINK_STATE_OFF &&
           data->next_toggle > 0u &&
           data->now >= data->next_toggle;
}

static void enter_blink_x1(rx_fsm_context *ctx, void *user) {
    blink_machine_data *data = (blink_machine_data *)user;

    (void)ctx;
    if (data == NULL) {
        return;
    }

    data->event_consumed = true;
    data->output_enabled = 1;
    data->next_toggle = rx_tick_add_us(data->now, half_period_us_for_state(BLINK_STATE_X1, data->base_hz));
}

static void enter_blink_x2(rx_fsm_context *ctx, void *user) {
    blink_machine_data *data = (blink_machine_data *)user;

    (void)ctx;
    if (data == NULL) {
        return;
    }

    data->event_consumed = true;
    data->output_enabled = 1;
    data->next_toggle = rx_tick_add_us(data->now, half_period_us_for_state(BLINK_STATE_X1, data->base_hz));
}

static void toggle_light(rx_fsm_context *ctx, void *user) {
    blink_machine_data *data = (blink_machine_data *)user;
    uint64_t half_period_us;

    (void)ctx;
    if (data == NULL) {
        return;
    }

    data->output_enabled = !data->output_enabled;
    half_period_us = half_period_us_for_state(data->machine->state, data->base_hz);
    if (data->next_toggle == 0u) {
        data->next_toggle = rx_tick_add_us(data->now, half_period_us);
        return;
    }
    data->next_toggle = rx_tick_add_us(data->next_toggle, half_period_us);
}

static void enter_off(rx_fsm_context *ctx, void *user) {
    blink_machine_data *data = (blink_machine_data *)user;

    (void)ctx;
    if (data == NULL) {
        return;
    }

    data->event_consumed = true;
    data->output_enabled = 0;
    data->next_toggle = 0u;
}

static void blink_machine_dump_outputs(rx_fsm_context *ctx, void *user) {
    blink_machine_data *data = (blink_machine_data *)user;

    (void)ctx;
    if (data == NULL) {
        return;
    }

    app_driver_set_light(data->light_gpio, data->output_enabled);
    if (data->event_consumed) {
        app_driver_clear_button_event(data->button_gpio);
    }
}

void blink_fsm_create(
    rx_fsm_machine *machine,
    gpio_num_t button_gpio,
    gpio_num_t light_gpio,
    unsigned int base_hz
) {
    static const rx_fsm_transition transitions[] = {
        {BLINK_STATE_OFF, BLINK_STATE_X1, button_pressed, enter_blink_x1},
        {BLINK_STATE_X1, BLINK_STATE_X2, button_pressed, enter_blink_x2},
        {BLINK_STATE_X1, BLINK_STATE_X1, toggle_due, toggle_light},
        {BLINK_STATE_X2, BLINK_STATE_OFF, button_pressed, enter_off},
        {BLINK_STATE_X2, BLINK_STATE_X2, toggle_due, toggle_light},
    };
    blink_machine_data *data;

    if (machine == NULL || base_hz == 0u || s_count >= RXNET_MAX_RUNTIME_NODES) {
        return;
    }
    data = &s_data[s_count++];

    data->machine = machine;
    data->button_gpio = button_gpio;
    data->light_gpio = light_gpio;
    data->latched_event = false;
    data->event_consumed = false;
    data->output_enabled = 0;
    data->base_hz = base_hz;
    data->now = 0;
    data->next_toggle = 0;

    if (app_driver_init_button(button_gpio) != ESP_OK ||
        app_driver_init_light(light_gpio) != ESP_OK) {
        --s_count;
        return;
    }

    rx_fsm_machine_init(
        machine,
        "blink",
        BLINK_STATE_OFF,
        transitions,
        sizeof(transitions) / sizeof(transitions[0]),
        data,
        blink_machine_latch_inputs,
        blink_machine_dump_outputs
    );
}

int blink_fsm_set_base_hz(rx_fsm_machine *machine, unsigned int base_hz) {
    blink_machine_data *data;

    if (machine == NULL || machine->user == NULL || base_hz == 0u) {
        return -1;
    }
    data = (blink_machine_data *)machine->user;
    data->base_hz = base_hz;
    if (machine->state != BLINK_STATE_OFF) {
        data->next_toggle = rx_tick_add_us(data->now, half_period_us_for_state(data->machine->state, data->base_hz));
    }
    return 0;
}

unsigned int blink_fsm_get_base_hz(const rx_fsm_machine *machine) {
    if (machine == NULL || machine->user == NULL) {
        return 0u;
    }
    return ((const blink_machine_data *)machine->user)->base_hz;
}

int blink_fsm_get_output_enabled(const rx_fsm_machine *machine) {
    if (machine == NULL || machine->user == NULL) {
        return 0;
    }
    return ((const blink_machine_data *)machine->user)->output_enabled;
}

File: python/examples/fsm/03-blink/blink_fsm.py
# Copyright 2026 Jose M. Moya <jm.moya@upm.es>
# SPDX-License-Identifier: GPL-3.0-or-later

"""Blink FSM — three-state blinking light.

States: OFF → X1 (blink at base_hz) → X2 (blink at 2×base_hz) → OFF.
Button press cycles through states.  A self-loop transition fires on each
half-period to toggle the physical output.
Equivalent to ``c/examples/fsm/02-blink/blink_fsm.c``.
"""
from __future__ import annotations

import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[4]))
sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from rxnet.fsm import Machine, Transition
from rxnet.runtime import Context

import app_driver

BLINK_STATE_OFF = 0
BLINK_STATE_X1 = 1
BLINK_STATE_X2 = 2


class _Data:
    __slots__ = (
        "button_gpio", "light_gpio",
        "latched_event", "event_consumed", "output_enabled",
        "base_hz", "now_ms", "next_toggle_ms",
    )

    def __init__(self, button_gpio: int, light_gpio: int, base_hz: int) -> None:
        self.button_gpio = button_gpio
        self.light_gpio = light_gpio
        self.latched_event = False
        self.event_consumed = False
        self.output_enabled = False
        self.base_hz = base_hz
        self.now_ms = 0
        self.next_toggle_ms = 0


def _now_ms() -> int:
    return int(time.monotonic() * 1000)


def _half_period_ms(base_hz: int, double_speed: bool) -> int:
    hz = max(base_hz, 1)
    if double_speed:
        hz *= 2
    hp = 500 // hz
    return max(hp, 1)


def _latch(ctx: Context, data: _Data) -> None:
    data.now_ms = _now_ms()
    data.latched_event = app_driver.latch_button_event(data.button_gpio)
    data.event_consumed = False


def _button_pressed(ctx: Context, data: _Data) -> bool:
    return data.latched_event


def _toggle_due(ctx: Context, data: _Data) -> bool:
    return (
        data.next_toggle_ms > 0
        and data.now_ms >= data.next_toggle_ms
    )


def _enter_x1(ctx: Context, data: _Data) -> None:
    data.event_consumed = True
    data.output_enabled = True
    data.next_toggle_ms = data.now_ms + _half_period_ms(data.base_hz, False)


def _enter_x2(ctx: Context, data: _Data) -> None:
    data.event_consumed = True
    data.output_enabled = True
    data.next_toggle_ms = data.now_ms + _half_period_ms(data.base_hz, True)


def _enter_off(ctx: Context, data: _Data) -> None:
    data.event_consumed = True
    data.output_enabled = False
    data.next_toggle_ms = 0


def _toggle_light(ctx: Context, data: _Data) -> None:
    data.output_enabled = not data.output_enabled
    # Determine speed from the machine's committed state (available after commit,
    # because actions run as deferred actions after commit).
    # We peek at the machine via a closure — here data.machine is set after creation.
    machine: Machine = data._machine  # type: ignore[attr-defined]
    double_speed = machine.state == BLINK_STATE_X2
    hp = _half_period_ms(data.base_hz, double_speed)
    if data.next_toggle_ms == 0:
        data.next_toggle_ms = data.now_ms + hp
    else:
        data.next_toggle_ms += hp


def _dump(ctx: Context, data: _Data) -> None:
    app_driver.set_light(data.light_gpio, data.output_enabled)
    if data.event_consumed:
        app_driver.clear_button_event(data.button_gpio)


_TRANSITIONS = [
    Transition(BLINK_STATE_OFF, BLINK_STATE_X1, guard=_button_pressed, action=_enter_x1),
    Transition(BLINK_STATE_X1, BLINK_STATE_X2, guard=_button_pressed, action=_enter_x2),
    Transition(BLINK_STATE_X1, BLINK_STATE_X1, guard=_toggle_due, action=_toggle_light),
    Transition(BLINK_STATE_X2, BLINK_STATE_OFF, guard=_button_pressed, action=_enter_off),
    Transition(BLINK_STATE_X2, BLINK_STATE_X2, guard=_toggle_due, action=_toggle_light),
]


def create_blink_fsm(button_gpio: int, light_gpio: int, base_hz: int) -> Machine:
    app_driver.init_button(button_gpio)
    app_driver.init_light(light_gpio)
    data = _Data(button_gpio, light_gpio, base_hz)
    machine = Machine(
        name="blink",
        state=BLINK_STATE_OFF,
        transitions=_TRANSITIONS,
        user=data,
        latch_inputs_cb=_latch,
        dump_outputs_cb=_dump,
    )
    data._machine = machine  # type: ignore[attr-defined]
    return machine


def set_base_hz(machine: Machine, base_hz: int) -> bool:
    if not isinstance(machine.user, _Data) or base_hz <= 0:
        return False
    machine.user.base_hz = base_hz
    if machine.state != BLINK_STATE_OFF:
        double_speed = machine.state == BLINK_STATE_X2
        machine.user.next_toggle_ms = _now_ms() + _half_period_ms(base_hz, double_speed)
    return True


def get_base_hz(machine: Machine) -> int:
    if not isinstance(machine.user, _Data):
        return 0
    return machine.user.base_hz


def get_output_enabled(machine: Machine) -> bool:
    if not isinstance(machine.user, _Data):
        return False
    return machine.user.output_enabled
